### DISCUSSION TITLE
What are the best practices for mimir multi data center monitoring?

### DISCUSSION THREAD

**Discussion by shichanglin5** [grafana/mimir#6604] [Help and support]
Our production environment has 3 data centers (IDC) to ensure the external availability of our services

I know that mimir supports zone awareness. Based on cross-zone replicas, it ensures that when an IDC cannot be accessed, the entire mimir cluster is still available (if 2 of 3 regions are available, the mimir cluster is available)

**However, I feel that this cross-zone replica deployment (each IDC corresponds to one availability zone) is not suitable for our production environment, based on the following considerations:**
1. Cross-zone copy involves data transmission across the IDC network, the delay is relatively high, and it is easily affected by the network dedicated line (the scale of our monitoring timing is about 200 Mil)
2. Since Mimir writes monitoring data to all ingesters in the availability zone, the failure of any ingester will cause the zone where it is located to become unavailable. Once one ingester fails in each of the two zones, the entire cluster will become unavailable (no matter how many replicas) , so although cross-region replicas introduce zone-awareness, they also reduce the fault tolerance of the entire cluster.
3. **When an IDC cannot be accessed for some reason (such as a dedicated network line being disconnected), although the overall mimir cluster is still available, it is unavailable for this IDC (in this scenario, due to the The monitoring data of the services in the IDC cannot be successfully written to mimir. When the fault is restored, the monitoring data of the services running in this IDC will be missing, which will have a great impact on troubleshooting and analysis)**

### Here's the deployment plan I'm thinking of:
Each IDC deploys a 3-replica mimir cluster to form 3 independent mimir clusters (without distinguishing between availability zones). Each mimir cluster is only responsible for storing the monitoring data of the current IDC.

The disadvantage of this deployment method is that when an IDC network failure cannot be accessed by other IDCs, the monitoring data cannot be accessed by other IDCs. **However, this is acceptable. When the network is restored, the monitoring data will be complete.**

like this:
<img width="987" alt="image" src="https://github.com/grafana/mimir/assets/48707349/1c28f2fe-350d-42da-a341-022c1c3f2331">

**However, I encountered a problem: Since there are three independent mimir clusters, the data between the clusters cannot be well aggregated and returned to a unified view. Is there any good way to solve this problem, or is there any other deployment plan?**

**Comment by dimitarvdimitrov** [comment]
Thanks for providing this level of detail. Your observations are correct regarding the network traffic and availability requirements. 

However, in your multi-cluster architecture sharing memberlist (hence the hash ring too) will effectively mean that the mimir cluster is unified. Since you want to keep the 3 IDCs isolated, then you should run a unique memberlist for each IDC. 

One OSS tool to help with aggregating the query results from the 3 clusters is https://github.com/jacksontj/promxy. I don't have experience with it, but it claims to solve the problem of multiple clusters.

**Reply by shichanglin5** [reply]
"However, in your multi-cluster architecture sharing memberlist (hence the hash ring too) will effectively mean that the mimir cluster is unified. "

===>
something to note here is that although the three mimir clusters share the same memberlist, the configurations are different (different prefixes will be configured for the hash ring to isolate the hash rings of different  IDC-clusters)
at ` vendor/github.com/grafana/dskit/kv/client.go:56`:
```go
// Config is config for a KVStore currently used by ring and HA tracker,
// where store can be consul or inmemory.
type Config struct {
	Store       string `yaml:"store"`
	Prefix      string `yaml:"prefix" category:"advanced"`
	StoreConfig `yaml:",inline"`

	Mock Client `yaml:"-"`
}
```

**Reply by shichanglin5** [reply]
In this way, if there are no other problems, each IDC's mimir cluster will only focus on the corresponding hash ring (**rings with corresponding IDC prefix**, when the distributor writes data to the ingester, it will get the ingester ring formed by the ingesters deployed in the current IDC based on the current IDC prefix), thus realizing multiple mimirs The cluster shares the same member list, but the write paths are isolated

**Reply by shichanglin5** [reply]
Based on the shared memberlist, query can easily discover the ingester rings of multiple mimir clusters (for example, the monitoring in the program can detect the ingester rings of different mimir clusters based on prefix monitoring. In this way, when querying, it can aggregate multiple different mimir clusters. Clustered ingester data)
at `vendor/github.com/grafana/dskit/kv/memberlist/memberlist_client.go:101`
```go
// WatchPrefix calls f whenever any value stored under prefix changes.
// Part of kv.Client interface.
func (c *Client) WatchPrefix(ctx context.Context, prefix string, f func(string, interface{}) bool) {
	err := c.awaitKVRunningOrStopping(ctx)
	if err != nil {
		return
	}

	c.kv.WatchPrefix(ctx, prefix, c.codec, f)
}
```

**Reply by shichanglin5** [reply]
However, the current ring used in the query is fixed and can only be the ring of the current cluster ingester.
at `pkg/distributor/query.go:102`
```go
// GetIngesters returns a replication set including all ingesters.
func (d *Distributor) GetIngesters(ctx context.Context) (ring.ReplicationSet, error) {
	userID, err := tenant.TenantID(ctx)
	if err != nil {
		return ring.ReplicationSet{}, err
	}

	// If tenant uses shuffle sharding, we should only query ingesters which are
	// part of the tenant's subring.
	shardSize := d.limits.IngestionTenantShardSize(userID)
	lookbackPeriod := d.cfg.ShuffleShardingLookbackPeriod

	if shardSize > 0 && lookbackPeriod > 0 {
		return d.ingestersRing.ShuffleShardWithLookback(userID, shardSize, lookbackPeriod, time.Now()).GetReplicationSetForOperation(readNoExtend)
	}

	return d.ingestersRing.GetReplicationSetForOperation(readNoExtend)
}
```

Aggregating the ingester ring of multiple mimir clusters may require the introduction of a multiRing-like structure
```go
type multiRing struct{
     rings map[string]ReadRing // ring by idc
}
```
That's all I can think of right now

**Reply by shichanglin5** [reply]
Based on the above, any suggestions? Or can this function become a feature added to the roadmap?

**Reply by dimitarvdimitrov** [reply]
The multiRing will probably be the first step to this, but the querier will also need to issue two distinct queries to the two DCs and merge the results. Perhaps here we need to issue another concurrent request for the other ring and merge the two `results` together

https://github.com/grafana/mimir/blob/b34913fba0267cd8e2920dbebadba945bd397343/pkg/querier/distributor_queryable.go#L124-L127

---

**However**, I'm not entirely sure if querying multiple clusters is in Mimir's scope. I think it helps if you think about Mimir as a single self-sufficient cluster, as opposed to a mesh of clusters. Anything that needs to aggregate across clusters might be better off as a separate tool. 

Maybe someone like @zhehao-grafana might have thoughts about this.

**Reply by zhehao-grafana** [reply]
@shichanglin5 Thanks for raising the issue to us! currently, we don't have a plan to support multi-cluster and we still recommend using https://github.com/jacksontj/promxy for now. In the meanwhile, we are making changes internally to support multi-cluster and planning to have it in the coming year

**Comment by shichanglin5** [comment]
https://github.com/jacksontj/promxy is a good project, but I am worried about compatibility issues in the future and the inability to make good use of mimir's features, such as caching, current limiting, etc.

**Reply by dimitarvdimitrov** [reply]
each mimir cluster will do limiting and caching on its own if you point promxy to the query-frontend (or the reverse proxy in front of the QF). I'm not sure how much overhead there is in promxy's merging of different clusters. Overall, I agree it will be nicer if mimir could natively support multi-dc architectures without the networking & availability limitations.
